gsy_framework/validators/heat_pump_validator.py

```diff
@@ -0,0 +1,117 @@
+        if kwargs["consumption_profile_uuid"] is None:
+            cls._check_range(
+                name="consumption_kW", value=kwargs["consumption_kW"],
+                _min=HeatPumpSettings.MAX_POWER_RATING_KW_LIMIT.min,
+                _max=kwargs["maximum_power_rating_kW"])
+
+    @classmethod
+    def _validate_temp(cls, **kwargs):
+        """Validate temperature related arguments."""
+        temperature_arg_names = [
+            "min_temp_C", "max_temp_C", "initial_temp_C"]
+        for temperature_arg_name in temperature_arg_names:
+            cls._check_range(
+                name=temperature_arg_name, value=kwargs[temperature_arg_name],
+                _min=HeatPumpSettings.TEMP_C_LIMIT.min,
+                _max=HeatPumpSettings.TEMP_C_LIMIT.max)
+
+        min_temp_c = kwargs["min_temp_C"]
+        max_temp_c = kwargs["max_temp_C"]
+
+        if not min_temp_c <= max_temp_c:
+            raise GSyDeviceException(
+                {"misconfiguration": [
+                    "Requirement 'min_temp_C <= max_temp_C' is not met."]})
+
+        if (kwargs["external_temp_C"] is not None and
+                kwargs["external_temp_profile_uuid"] is not None):
+            raise GSyDeviceException(
+                {"misconfiguration": [
+                    "external_temp_C shouldn't be set with external_temp_profile_uuid."]})
+
+        if (kwargs["external_temp_C"] is None and
+                kwargs["external_temp_profile_uuid"] is None):
+            raise GSyDeviceException(
+                {"misconfiguration": [
+                    "Either external_temp_C or external_temp_profile_uuid should be set."]})
+
+        if kwargs["external_temp_profile_uuid"] is None:
+            cls._check_range(
+                name="external_temp_C", value=kwargs["external_temp_C"],
+                _min=HeatPumpSettings.TEMP_C_LIMIT.min,
+                _max=HeatPumpSettings.TEMP_C_LIMIT.max)
+
+    @classmethod
+    def _validate_rate(cls, **kwargs):
+        """Validate energy rate related arguments."""
+        buying_rate_arg_names = [
+            "initial_buying_rate", "preferred_buying_rate", "final_buying_rate"]
+        for buying_rate_arg_name in buying_rate_arg_names:
+            cls._check_range(
+                name=buying_rate_arg_name, value=kwargs[buying_rate_arg_name],
+                _min=HeatPumpSettings.TEMP_C_LIMIT.min,
+                _max=HeatPumpSettings.TEMP_C_LIMIT.max)
+
+        initial_buying_rate = kwargs["initial_buying_rate"]
+        preferred_buying_rate = kwargs["preferred_buying_rate"]
+        final_buying_rate = kwargs["final_buying_rate"]
+
+        validate_range_limit(
+            initial_buying_rate, preferred_buying_rate, final_buying_rate,
+            {"misconfiguration": [
+                "Requirement "
+                "'initial_buying_rate <= preferred_buying_rate <= final_buying_rate' is not met."
+            ]})
+
+    @classmethod
+    def _check_range(cls, name, value, _min, _max):
```

Comment:
Why did you create this wrapper around `validate_range_limit` and why are the input arguments prepended with `_`?

Reply:
I felt bad about copy-pasting the same code for populating the `error_message` everytime I wanted to validate something. At first, I tried to refactor validate_range_limit but then I saw it's heavily used in many places. So, I created my little wrapper here. Also, I couldn't use `min` and `max` as my arguments as they will shadow the built-in min and max functions of Python. I can rename it to `min_value` and `max_value` if you feel uncomforable.

Reply:
I would actually like the latter, because all arguments that begin with _ look like unused arguments to my eyes, and seeing them being used below confuses me a bit.

Reply:
Done.